gitpod-io/gitpod
# Opportunistic token refresh (for BitBucket Server)

## Description
This aims to introduce a new feature for token refreshes with SCMs dubbed "opportunistic refresh". If the feature flag is enabled, AND the SCM requires it (for now only BitBucket Server!), we ask for a fresh token whenever we "safely" can.

This "safely" is determined by us storing a new `reservedUntilDate` with each SCM token, with is set/updated whenever we re-use a token. This way, whenever we do a lookup on a SCM token, we know whether we can safely refresh it or not.

**Operational controls**:
 - introduces the `workspace_start_scm_access_token_lifetime` feature flag which controls the requested lifetime of a SCM token
   - the default for SCM tokens is 5mins
   - the default for SCM tokens on workspace start is 10mins
   - we can override the later with this feature flag
 - introduces the `opportunistic_token_refresh` feature flag which controls whether "opportunistic refresh" is enabled at all
 - updates the `gitpod_scm_token_refresh_requests_total` metric with a new label `opportunisticRefresh` (`"true" | "false" | "reserved"`)

## Related Issue(s)

Fixes ENT-4

## How to test
 - check the unit tests [here](https://github.com/gitpod-io/gitpod/pull/19715/files#diff-68da8a99b384ed20bcbaf557f87a2c9c0827fe138d7c158c7f4ec91c7ce51ccdR218-R313)
 - manual testing:
   - observe `kubectl logs server-5d5cdb8db5-tj9bl -f | grep "Token refreshed"` and noticed how values change
   - [join my org](https://gpl-bbs-toc9354957bc.preview.gitpod-dev.com/orgs/join?inviteId=80bd1d8e-cbc7-420a-94af-2f011b91becc), and open a repo from our BitBucket instance

## Documentation


#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

      Run the build with werft instead of GHA
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

</details>

<details>
<summary>Installer</summary>

  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

      If enabled this will build `install/preview`
      If enabled this will create the environment on GCE infra
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
</details>

/hold
